src/java.base/share/classes/jdk/internal/misc/CarrierThread.java

```diff
@@ -60,27 +66,43 @@ public CarrierThread(ForkJoinPool pool) {
     }
 
     /**
-     * For use by {@link Blocker} to test if the thread is in a blocking operation.
+     * Mark the start of a blocking operation.
      */
-    boolean inBlocking() {
-        //assert JLA.currentCarrierThread() == this;
-        return blocking;
-    }
+    public void beginBlocking() {
+        assert Thread.currentThread().isVirtual() && JLA.currentCarrierThread() == this;
+        if (compensating == NOT_COMPENSATING) {
+            compensating = COMPENSATE_IN_PROGRESS;
 
-    /**
-     * For use by {@link Blocker} to mark the start of a blocking operation.
-     */
-    void beginBlocking() {
-        //assert JLA.currentCarrierThread() == this && !blocking;
-        blocking = true;
+            // FJP.tryCompensate to create or re-activate a spare thread
+            long value;
+            try {
+                value = ForkJoinPools.beginCompensatedBlock(getPool());
+            } catch (Throwable e) {
+                if (compensating == COMPENSATE_IN_PROGRESS) {
+                    compensating = NOT_COMPENSATING;
+                }
+                throw e;
+            }
+
+            if (compensating == COMPENSATE_IN_PROGRESS) {
+                compensateValue = value;
+                compensating = COMPENSATING;
+            } else {
+                // forced preemption or preempted when attempting to compensate
+                ForkJoinPools.endCompensatedBlock(getPool(), value);
+            }
+        }
     }
 
     /**
-     * For use by {@link Blocker} to mark the end of a blocking operation.
+     * Mark the end of a blocking operation.
      */
-    void endBlocking() {
-        //assert JLA.currentCarrierThread() == this && blocking;
-        blocking = false;
+    public void endBlocking() {
+        assert Thread.currentThread() == this || JLA.currentCarrierThread() == this;
+        if (compensating == COMPENSATING) {
+            ForkJoinPools.endCompensatedBlock(getPool(), compensateValue);
+            compensating = NOT_COMPENSATING;
```

Comment:
For the sake for consistency between the state and the `compensateValue`, would it be good to reset `compensateValue` to `0` when we switch to `NOT_COMPENSATING`?